shared-types.d.ts

```diff
@@ -0,0 +1,274 @@
+// Minimum TypeScript Version: 3.0
+
+import Codes from './lib/character/codes'
+import Types from './lib/constant/types'
+
+/**
+ * A location in a string or buffer
+ */
+export interface Point {
+  line: number
+  column: number
+  offset: number
+  _index?: number
+  _bufferIndex?: number
+}
+
+/**
+ * A token type
+ */
+export type Type = Types[keyof Types]
```

Comment:
The goal is also to have types extendable btw, e.g., `micromark-extension-gfm` adds other types.

Would that work with this?

Reply:
I will do some research on extendable types.

Reply:
> From these three points, could we have the source in lib/, copy them into dist/ when building, and then in package.json ignore the ones in lib/ but publish the ones in dist/?

This sounds ok to me. I will do this later.

Reply:
Here are the ones for mdast and hast, which might serve as inspiration: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mdast/index.d.ts
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/hast/index.d.ts

It doesn’t look like they do anything for extending tho

Reply:
Maybe we can make `buffer` a function with two generic types: `SE extends SyntaxExtension`, `EE extends HtmlExtension` so that we can pass the extra `Type` into it.

something like this:

```
declare function buffer<
  SE extends SyntaxExtension,
  EE extends HtmlExtension,
  >(
    value: string | Buffer,
    encoding?: BufferEncoding,
    options?: ParseOptions<SE, EE>
  ): string
```
 

But this is too much modification: every function that accepts `Token` will need to add generic. So for now, let's just set  `Type` as `string` to avoid errors.

Reply:
Sounds good to me! @ChristianMurphy might have ideas though

Reply:
:thinking: I don't think we'll ever remove types? Only add them?
If so, something like an interestion `type Type<Extras> = Types[keyof Types] & Extras`.
Then as @ocavue suggests those types would likely bubble up to higher level types.

Reply:
@ChristianMurphy Do you think it's OK to just keep the way below in this PR? Since it's more simple.

```
export type Type = string
export type Code = number | null
```

Reply:
:thinking: that may be fine, it can be revisited if any of the extensions need a more specific type.

---

The principle I try to follow with typings is "Make invalid states unrepresentable".
Meaning not only should valid micromark calls have types be allowed, but also invalid calls to micromark should, as much as TypeScript allows, throw an error.
As long as we aren't introducing potentially invalid program states to the typings, I am all for simpler representations :+1: